### ISSUE TITLE
Tape-recorded `wp.copy()` / `wp.clone()` / `array.assign()` produces silently wrong gradients

### ISSUE THREAD

**Issue by daedalus5** [NVIDIA/warp#1728]
### Bug Description

The adjoint of a tape-recorded copy violated Warp's adjoint conventions in both
directions, producing silently wrong gradients with no warning (the forward pass
contains no write-after-read, so the overwrite detector is blind to it):

1. **Source: assignment instead of accumulation.** When the copy's source is also read
   by a kernel launched later in the forward pass, that kernel's adjoint contribution
   lands in the source gradient first (backward runs in reverse) and the copy's adjoint
   overwrote it. In the common iterative pattern the gradients come out exactly zero.
2. **Destination: adjoint never consumed.** The copy overwrites its destination, so any
   earlier write to that region is dead under final-write-wins — yet its adjoint leaked,
   propagating phantom gradients through dead writes. Kernel adjoints zero their consumed
   output gradients; the copy adjoint now matches (`retain_grad=True` destinations exempt).

```python
@wp.kernel
def mul_constant(x: wp.array[float], y: wp.array[float]):
    tid = wp.tid()
    y[tid] = x[tid] * 2.0

@wp.kernel
def sum_kernel(x: wp.array[float], total: wp.array[float]):
    tid = wp.tid()
    wp.atomic_add(total, 0, x[tid])

n = 4
x = wp.array(np.arange(1.0, n + 1), dtype=float, requires_grad=True)
loss = wp.zeros(1, dtype=float, requires_grad=True)

tape = wp.Tape()
with tape:
    q = x
    for _ in range(3):
        out = wp.clone(q)                                  # q is read by the clone...
        wp.launch(mul_constant, dim=n, inputs=[q, out])    # ...and by this later kernel
        q = out
    wp.launch(sum_kernel, dim=n, inputs=[q], outputs=[loss])
tape.backward(loss)

print(x.grad.numpy())  # unfixed: [0. 0. 0. 0.] — expected [8. 8. 8. 8.] (2^3)
```

With partial writes instead of full overwrites the gradients are wrong by a large
data-dependent factor rather than zero (verified 45–90% of true values in VBD-style
clone-plus-partial-write patterns).

**Fix:** extract the incoming adjoint into a zero-initialized temporary (reusing
`copy()`'s offset/layout handling), zero the destination region's adjoint via the
mirrored copy call, and add the extracted adjoint elementwise into the source gradient;
synchronize the copy stream first when the addition can't be stream-ordered
(device-to-host and cross-device copies). Tested for 1–4D arrays, vector dtypes,
float64, non-contiguous sources, offset ranges, cpu↔cuda copies, dead-write consumption
(full/partial/repeated), and the `retain_grad` exemption.

**Comment by shi-eric** [comment]
Progress update: 4cb14232712edb6a69e8e7dffaa7fc4da68c3ad2 landed a scoped fix for tape-recorded same-device copies.

The change fixes source-adjoint accumulation and destination-adjoint consumption for same-device, non-overlapping copies of matching addable numeric dtypes. This covers the reported iterative clone pattern, partial overwrites, `array.assign()`, and the `retain_grad=True` exemption.

The issue explicitly includes CPU↔CUDA copies, which remain on the previous byte-copy adjoint path and still produce incorrect gradients. The issue should therefore remain open.

### Public API behavior

- Existing `wp.copy()`, `wp.clone()`, and `array.assign()` operations now accumulate into the source gradient and consume the overwritten destination gradient for supported same-device copies.
- Cross-device copies, overlapping regions, struct and boolean dtypes, reinterpreting dtypes, non-positive-stride regions, and streams belonging to another device retain the previous adjoint behavior.

### Test coverage

- `warp/tests/test_tape.py::test_tape_copy_adjoint_accumulation` covers direct and iterative accumulation, `array.assign()`, and equivalent vector dtype spellings.
- `warp/tests/test_tape.py::test_tape_copy_adjoint_consumption` covers full, partial, and repeated dead writes plus the `retain_grad=True` exemption.
- `warp/tests/test_tape.py::test_tape_copy_adjoint_views_and_offsets` covers offset windows, strided views, non-contiguous destinations, and 3D `float64` views.
- `warp/tests/test_tape.py::test_tape_copy_adjoint_out_of_scope` pins the previous behavior for overlapping, interleaved, struct, boolean, mixed-dtype, and cross-device copies.
- `warp/tests/test_tape.py::test_tape_copy_adjoint_stream` and `test_tape_copy_adjoint_graph_capture` cover non-current streams and CUDA graph replay.
- No committed 4D regression was added.

### Behavioral verification

Standalone probes reproduced the original issue on both CPU and CUDA and confirmed the expected `[8, 8, 8, 8]` gradient after the fix. Same-device partial-overwrite, dead-write-consumption, and 4D cases also produced their analytic gradients.

CPU→CUDA and CUDA→CPU probes still produced `[1, 1, 1, 1]` where the analytic gradient is `[3, 3, 3, 3]`.

### Follow-up to consider

- Extend correct adjoint accumulation and destination consumption to CPU↔CUDA copies; this is the remaining closure blocker.
- Add a committed 4D regression test. Standalone 4D behavior passes, so this is a coverage improvement rather than an additional behavioral blocker.

**Comment by shi-eric** [comment]
This is addressed by dc20e08c72de1f70ad05a05d9d8442236c62a27b.

The change completes the remaining CPU↔CUDA portion of this issue for contiguous copies of matching addable numeric dtypes. Tape backward now accumulates the incoming destination adjoint into the source gradient and consumes the overwritten destination gradient unless `retain_grad=True`. Together with the same-device behavior documented above, this covers the reported source-accumulation and dead-write-consumption failures.

Unsupported copy-adjoint patterns now emit a `UserWarning` explaining that gradients cannot be fully tracked instead of silently using the fallback behavior.

### Public API behavior

- Existing tape-recorded `wp.copy()`, `wp.clone()`, and `array.assign()` operations now produce correct source accumulation and destination consumption for supported same-device and contiguous CPU↔CUDA copies.
- Partial ranges and `retain_grad=True` are supported on the corrected paths.
- CPU↔CUDA copy adjoints explicitly raise during CUDA graph capture.
- CUDA peer copies, direct non-contiguous CPU/CUDA adjoint fallbacks, unsupported or reinterpreted gradient dtypes, overlapping copies, and streams from another device retain fallback behavior and issue a warning; correct gradients are not guaranteed for those patterns.

### Test coverage

- New `warp/tests/test_tape.py::test_tape_copy_adjoint_cpu_cuda` covers source accumulation and destination consumption in both CPU↔CUDA directions, an offset range on a non-current stream, and `retain_grad=True`.
- New `warp/tests/test_tape.py::test_tape_copy_adjoint_fallback_warning` checks warning reasons for overlapping copies, non-contiguous CPU/CUDA fallbacks, and incompatible gradient dtypes.
- `warp/tests/test_tape.py::test_tape_copy_adjoint_views_and_offsets` now includes a 4D accumulation-and-consumption regression.
- `warp/tests/test_tape.py::test_tape_copy_adjoint_graph_capture` now checks that both CPU→CUDA and CUDA→CPU adjoints reject CUDA graph capture.
- `warp/tests/test_tape.py::test_tape_copy_adjoint_out_of_scope` was updated to account for warnings on fallback paths.

### Behavioral verification

Standalone probes reproduced the original iterative clone case on CPU and CUDA and confirmed the expected `[8, 8, 8, 8]` source gradient. Both CPU↔CUDA directions also produced the expected gradients through `wp.copy()`, `wp.clone()`, and `array.assign()`, including partial ranges, `retain_grad`, 4D `float64`, vector dtypes, and non-contiguous sources.

### Follow-up to consider

- Add committed CPU↔CUDA coverage for `wp.clone()`, `array.assign()`, multidimensional `float64`, and vector dtypes. Standalone probes pass, so this does not block closure.
- Extend correct destination-gradient consumption to CUDA→CPU copies into non-contiguous CPU destinations. This path currently warns and remains outside the supported adjoint behavior, so it does not block closure.
